services/monitor/src/ChainMonitor.ts

```diff
@@ -0,0 +1,281 @@
+import assert from "assert";
+import { EventEmitter } from "stream";
+import { decode as bytecodeDecode } from "@ethereum-sourcify/bytecode-utils";
+import { SourcifyChain } from "@ethereum-sourcify/lib-sourcify";
+import logger from "./logger";
+import { KnownDecentralizedStorageFetchers, MonitorConfig } from "./types";
+import { Logger } from "winston";
+import PendingContract from "./PendingContract";
+
+function createsContract(tx: TransactionResponse): boolean {
+  return !tx.to;
+}
+
+const NEW_BLOCK_EVENT = "new-block";
+
+/**
+ * A monitor that periodically checks for new contracts on a single chain.
+ */
+export class ChainMonitor extends EventEmitter {
+  private sourcifyChain: SourcifyChain;
+  private sourceFetchers: KnownDecentralizedStorageFetchers;
+  private sourcifyServerURLs: string[];
+
+  private chainLogger: Logger;
+  private startBlock?: number;
+  private blockInterval: number;
+  private blockIntervalFactor: number;
+  private blockIntervalUpperLimit: number;
+  private blockIntervalLowerLimit: number;
+  private bytecodeInterval: number;
+  private bytecodeNumberOfTries: number;
+
+  constructor(
+    sourcifyChain: SourcifyChain,
+    sourceFetchers: KnownDecentralizedStorageFetchers,
+    monitorConfig: MonitorConfig
+  ) {
+    super();
+    this.sourcifyChain = sourcifyChain;
+    this.sourceFetchers = sourceFetchers; // TODO: handle multipe
+    this.chainLogger = logger.child({
+      prefix: "Chain #" + this.sourcifyChain.chainId,
+    });
+
+    this.chainLogger.info(
+      `Initializing chain monitor for chain with sourceFetchers: ${Object.keys(
+        sourceFetchers
+      ).join(",")}`
+    );
+    this.sourcifyServerURLs = monitorConfig.sourcifyServerURLs;
+
+    const chainConfig = {
+      ...monitorConfig.defaultChainConfig,
+      ...(monitorConfig.chainConfigs?.[this.sourcifyChain.chainId] || {}),
+    };
+
+    this.startBlock = chainConfig.startBlock;
+    this.bytecodeInterval = chainConfig.bytecodeInterval;
+    this.blockInterval = chainConfig.blockInterval;
+    this.blockIntervalFactor = chainConfig.blockIntervalFactor;
+    assert(this.blockIntervalFactor > 1);
+    this.blockIntervalUpperLimit = chainConfig.blockIntervalUpperLimit;
+    this.blockIntervalLowerLimit = chainConfig.blockIntervalLowerLimit;
+    this.bytecodeNumberOfTries = chainConfig.bytecodeNumberOfTries;
+  }
+
+  start = async (): Promise<void> => {
+    try {
+      const lastBlockNumber = await this.sourcifyChain.getBlockNumber();
+      const startBlock =
+        this.startBlock !== undefined ? this.startBlock : lastBlockNumber;
+
+      this.chainLogger.info(`Starting monitor on block ${startBlock}`);
+
+      // Start polling
+      this.pollBlocks(startBlock);
+
+      // Listen to new blocks
+      this.on(NEW_BLOCK_EVENT, this.processBlockListener);
+    } catch (err: any) {
+      this.chainLogger.error(`Error starting monitor: ${err.message}`);
```

Comment:
Here we can add some information about the `ChainMonitor` (chainId?) so one knows which monitor failed directly from the error